HOLD - Do not merge - [Stream] Content reorg to align with PCX Content Strategy (#3221)
cloudflare/cloudflare-docs: products/stream/src/content/how-to/upload-videos/upload-video-file.md

---

> title: Upload video files

Reviewer: Removed as it's the same title as below

order: 2
pcx-content-type: how-to
---

# Upload video files

This option for uploading videos is ideal when the video is stored on a computer with access to your API tokens.

## Basic Uploads (small videos)

For files smaller than 200MB, you can use form based uploads, but this option does not support resumable uploading.

Make an HTTP request with content-type header set to `multipart/form-data` and include the media as an input with the name set to `file`.

```bash
curl -X POST \
-H "Authorization: Bearer $TOKEN" \
-F file=@/Users/kyle/Desktop/video.mp4 \
https://api.cloudflare.com/client/v4/accounts/$ACCOUNT/stream
```

<Aside>

Note that cURL `-F` flag automatically configures the content-type header and maps `skiing.mp4` to a form input called `file`.

</Aside>

## Resumable uploads with tus (large files)

> <details>

Reviewer: I find it a bit weird to have a "details" markup explanation here. I would just cut/paste this explanation and paste it below the paragraph explaining that tus is the recommended protocol for large files, as a regular paragrah. Wdyt?

<summary>
  What is tus?
</summary>
<div>
tus is a protocol based on HTTP for resumable file uploads. Resumable means that an upload can be interrupted at any moment and can be resumed without re-uploading the previous data again. An interruption may happen willingly, if the user wants to pause, or by accident in case of an network issue or server outage.
</div>
</details>

The [tus protocol](https://tus.io) is the recommended method for uploading large files to Cloudflare Stream from a computer, and popular programming languages have [tus client implementations](https://tus.io/implementations.html).

Cloudflare Stream requires a minimum chunk size of 5,242,880 bytes when using tus, unless the entire file is less than this amount.

We recommend increasing the chunk size to 52,428,800 bytes for better performance when the client connection is expected to be reliable. The maximum chunk size can be 209,715,200 bytes.

> <Aside>
> 
> Cloudflare Stream requires a chunk size divisible by 256KiB (256x1024 bytes). Round your desired chunk size to the nearest multiple of 256KiB.
> 
> The final chunk of an upload or uploads that fit within a single chunk are exempt from this requirement.
> 
> </Aside>

Reviewer: ```suggestion
<Aside type="note">

Cloudflare Stream requires a chunk size divisible by 256 KiB (256x1024 bytes). Round your desired chunk size to the nearest multiple of 256 KiB.

The final chunk of an upload or uploads that fit within a single chunk are exempt from this requirement.

</Aside>
```

## Specify upload options

The tus protocol allows you to add optional parameters in the [`Upload-Metadata` header](https://tus.io/protocols/resumable-upload.html#upload-metadata) and setting the arbitrary metadata values in the `Upload-Metadata` header sets values in the meta key in the Stream API. For a list of acceptable values, refer to the [Upload a video from a URL endpoint's optional parameters](https://api.cloudflare.com/#stream-videos-properties).

## Get the video ID using tus

> When an initial TUS request is made, Stream responds with a URL in the location header. While this URL may contain the video ID, we do not recommend parsing this URL to get the ID.

Reviewer: ```suggestion
When an initial tus request is made, Stream responds with a URL in the location header. While this URL may contain the video ID, we do not recommend parsing this URL to get the ID.
```

Reviewer: You've been using `tus` lowercase so far

Instead, use the `stream-media-id` HTTP header in the response to retrieve the video ID.

> For example, using the tus protocol to make a request to `https://api.cloudflare.com/client/v4/accounts/$ACCOUNT/stream`
>   
> will contain an HTTP header like the one below.

Reviewer: ```suggestion
For example, using the tus protocol to make a request to `https://api.cloudflare.com/client/v4/accounts/$ACCOUNT/stream` will contain an HTTP header like the one below.
```

```
stream-media-id: cab807e0c477d01baq20f66c3d1dfc26cf
```

### Command-line

You will also need to download a tus client. This example will use the [tus Python client](https://github.com/tus/tus-py-client), available through pip, Python's package manager.

```bash
pip install -U tus.py
```

```bash
tus-upload --chunk-size 52428800 --header Authorization "Bearer $TOKEN" $PATH_TO_VIDEO https://api.cloudflare.com/client/v4/accounts/$ACCOUNT/stream
```

At the beginning of the tus response, you will find the endpoint for getting information about your newly uploaded video.

    INFO Creating file endpoint
    INFO Created: https://api.cloudflare.com/client/v4/accounts/d467d4f0fcbcd9791b613bc3a9599cdc/stream/dd5d531a12de0c724bd1275a3b2bc9c6

>     ...

Reviewer: I think this should be placed between txt backticks but I don't know how to do it here with the Insert suggestion option

### Golang

Import a [tus](https://tus.io) client. You can use the [go-tus](https://github.com/eventials/go-tus) by [eventials](https://github.com/eventials) to upload from your Go applications.

```go
package main

import (
	"net/http"
	"os"

	tus "github.com/eventials/go-tus"
)

func main() {
	accountID := "ACCOUNT ID"

	f, err := os.Open("videofile.mp4")

	if err != nil {
		panic(err)
	}

	defer f.Close()

	headers := make(http.Header)
	headers.Add("Authorization", "Bearer $TOKEN")

	config := &tus.Config{
		ChunkSize:           50 * 1024 * 1024, // Required a minimum chunk size of 5MB, here we use 50MB.
		Resume:              false,
		OverridePatchMethod: false,
		Store:               nil,
		Header:              headers,
		HttpClient:          nil,
	}

	client, _ := tus.NewClient("https://api.cloudflare.com/client/v4/accounts/"+ accountID +"/stream", config)

	upload, _ := tus.NewUploadFromFile(f)

	uploader, _ := client.CreateUpload(upload)

	uploader.Upload()
}

```

> You can also get the progress of the upload if you' are running the upload in a goroutine.

Reviewer: ```suggestion
You can also get the progress of the upload if you are running the upload in a goroutine.
```

```go
// returns the progress percentage.
upload.Progress()

// returns whether or not the upload is complete.
upload.Finished()
```

Refer to [go-tus](https://github.com/eventials/go-tus) on GitHub for functionality such as resuming uploads and getting details about the progress of the upload.

### Node.js

1. Install `tus-js-client`.

```bash
npm install tus-js-client
```

2. Set up an `index.js` and configure:

> * The API endpoint with your Cloudflare Account ID
> * The request headers to include an API token

Reviewer: ```suggestion
* The API endpoint with your Cloudflare Account ID.
* The request headers to include an API token.
```
